### ISSUE TITLE
Arguments injection syntax

### ISSUE THREAD

**Issue by AleF83** [Soluto/stitch#149]
# Arguments Injection

Some of the directives has arguments to be set when attached to field. Setting static values is not enough. Sometimes the directive should get request JWT claim as argument, sometimes it can be value of parent object.

## Current state
In currently implemented solution we can evaluate expressions only in two ways.

1. String template with tokens what can be only property of one of the sources.
2. Value of any type that can be only property of one of the sources.

There is no support for any other operations (array index, math operations, string functions).

## Proposal

Argument injection mechanism will allow to use expressions in arguments. The expressions will be evaluated at request execution time.

### Sources

The argument expression can be dependent on several sources:

* Request JWT
* Field arguments
* Parent fields
* Ancestor's fields exported higher in query result tree
* *? GraphQL query variables ?*

#### Static value

The value is set hardcoded.

```graphql
type Query {
  user(id: ID!): User! @myDirective(arg1: "static")
```

It can be of any type.

```graphql
type Query {
  user(id: ID!): User! @myDirective(arg1: { a: 5 })
```

#### JavaScript Expression

The expression is padded with `"{` and `}"`. It is evaluated as JS expression. The argument value type will be the type of evaluated value.

```graphql
type Query {
  user(id: ID!): User! @myDirective(arg1: "{source.users[args.id]}")
```

#### String template

The expression is of type string and there are `${<expression>}` tokens inside. Each token is evaluated as JS expression (see above). The argument type will be always string.

```graphql
type Query {
  user(id: ID!): User! @myDirective(arg1: "UserId is ${args.id}")
```

> Notice: Be careful when the expression has only one token. `"{1+1}"` will be evaluated to `2`. But `"${1+1}"` will be evaluated to `"2"`.

**Comment by AleF83** [comment]
Actually there is no difference between `"something ${<expression>} something else"` and `"something {<expression>} something else}`. So we can just support the second option and the case of `"${<expression>}"` (just to cast value to string).

**Comment by AvivRubys** [comment]
I see two different features here:
1. Supporting single-template strings resolving to their correct type instead of string. So `{args.count}` resolving to `8` instead of `"8"`. This seems perfectly reasonable to me.
2. Letting users run arbitrary JS in templates. I'm worried this leads stitch into ESB-land, and thus I'm skeptical of this. What's the use case you have in mind for this?

**Comment by Yshayy** [comment]
The gql schema type will probably always be a string (if we want that all files will be compatible as gql files)
```
type Query {
  user(id: ID!): User! @myDirective(arg1: { a: 5 })
```
Is actually 
```
type Query {
  user(id: ID!): User! @myDirective(arg1: "{ a: 5 }")
```
arg1 is defined as a string, but resolves in runtime to an object

**Comment by Yshayy** [comment]
@AvivRubys I think it's easier and simple to use script injection than to create a value interpreter (even one that supports only dot/array notation, indexes, and string interpolation). It obviously comes with security risk, abuse risk and runtime exceptions, but I think these risks are minimal.
The syntax is regular javascript which is nice, in a similar fashion to firebase authorization rules or JSX/templating engine which makes it easy to explain and use.

**Comment by Yshayy** [comment]
How do we deal with null reference? currently and with the new approach?

**Comment by Yshayy** [comment]
Do we want to add backward compatibility (support for {} in templates) like in the gist example:
https://gist.github.com/Yshayy/9dda7f7b7dbd9dde355fb09468baecaa

**Comment by AleF83** [comment]
> The gql schema type will probably always be a string (if we want that all files will be compatible as gql files)
> 
> ```
> type Query {
>   user(id: ID!): User! @myDirective(arg1: { a: 5 })
> ```
> 
> Is actually
> 
> ```
> type Query {
>   user(id: ID!): User! @myDirective(arg1: "{ a: 5 }")
> ```
> 
> arg1 is defined as a string, but resolves in runtime to an object

This example is simple graphql argument definition. We cannot prevent user from do it. And shouldn't. The argument injection mechanism should be invoked only for argument of string type. I mentioned this option only for the full picture description.

**Comment by AleF83** [comment]
> How do we deal with null reference? currently and with the new approach?

Currently we don't deal with any exception that can be thrown during the argument evaluation.
It's a good question. Because the argument expression is defined by the schema writer but the exception will be thrown to user.
Maybe we can try to do some checks on schema mutation but I don't think we can cover all cases.
So maybe we can throw `GraphQLException` to signal to Apollo server to add it to `errors` property in response.

**Comment by AleF83** [comment]
> Do we want to add backward compatibility (support for {} in templates) like in the gist example:
> https://gist.github.com/Yshayy/9dda7f7b7dbd9dde355fb09468baecaa

We can just use `{<expression>}`. String template is a private case of expression. So it will be backward compatible.
See my first comment:

>Actually there is no difference between "something ${<expression>} something else" and "something {<expression>} something else}. So we can just support the second option and the case of "${<expression>}" (just to cast value to string).

**Comment by AleF83** [comment]
Implemented in #151
